flink-connectors/flink-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseSinkFunctionBase.java

```diff
@@ -0,0 +1,332 @@
+		this.columnFamilies = columnFamilies;
+		this.qualifiers = qualifiers;
+		this.fieldTypes = fieldTypes;
+
+		int[] fieldElementIndexMapping = new int[fieldNames.length];
+		for (int i = 0; i < fieldNames.length; i++) {
+			fieldElementIndexMapping[i] = -1;
+			for (int j = 0; j < outputFieldNames.length; j++) {
+				if (fieldNames[i].equals(outputFieldNames[j])) {
+					fieldElementIndexMapping[i] = j;
+					break;
+				}
+			}
+			if (fieldElementIndexMapping[i] == -1) {
+				throw new RuntimeException("The field " + outputFieldNames[i] + " is not found in the result stream.");
+			}
+		}
+		this.fieldElementIndexMapping = fieldElementIndexMapping;
+
+		batchFlushEnable = userConfig.getOrDefault(CONFIG_KEY_BATCH_FLUSH_ENABLE, "false").equals("true");
+		batchFlushMaxMutations = Long.parseLong(userConfig.getOrDefault(CONFIG_KEY_BATCH_FLUSH_MAX_MUTATIONS, "128"));
+		batchFlushMaxSizeInBits = Long.parseLong(userConfig.getOrDefault(CONFIG_KEY_BATCH_FLUSH_MAX_SIZE_MB, "2")) * 1024 * 1024 * 8;
+		batchFlushIntervalMillis = Long.parseLong(userConfig.getOrDefault(CONFIG_KEY_BATCH_FLUSH_INTERVAL_MS, "1000"));
+		batchFlushMaxRetries = Integer.parseInt(userConfig.getOrDefault(CONFIG_KEY_BATCH_FLUSH_MAX_RETRIES, "3"));
+		batchFlushMaxTimeoutMillis = Long.parseLong(userConfig.getOrDefault(CONFIG_KEY_BATCH_FLUSH_MAX_TIMEOUT_MS, "5000"));
+	}
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		this.lock = new Object();
+		if (isBuildingTable) {
+			if (client != null) {
+				this.connection = client.buildConnection();
+			}
+			if (connection != null) {
+				this.hTable = client.buildTable(connection);
+			}
+		}
+		if (hTable == null) {
+			throw new RuntimeException("Cannot build connection for hbase sink, please check the configuraiton.");
+		}
+		if (batchFlushEnable) {
+			((HTable) hTable).setAutoFlush(false, false);
+		} else {
+			((HTable) hTable).setAutoFlush(true, false);
+		}
+		this.executor = new ScheduledThreadPoolExecutor(1);
+		this.flushExecutor = Executors.newFixedThreadPool(3);
+		if (batchFlushEnable && batchFlushIntervalMillis > 0) {
+			executor.scheduleAtFixedRate(() -> {
+				if (this.hTable != null && this.hTable instanceof HTable) {
+					synchronized (lock) {
+						try {
+							flushToHBaseWithRetryAndTimeout();
+						} catch (Exception e){
+							log.warn("Scheduled flush operation to HBase cannot be finished.", e);
+						}
+					}
+				}
+			}, batchFlushIntervalMillis, batchFlushIntervalMillis, TimeUnit.MILLISECONDS);
+		}
+		isRunning = true;
+	}
+
+	@Override
+	public void invoke(IN value, Context context) throws Exception {
+		Mutation mutation = extract(value);
+		long mutationSize = mutation.heapSize();
+		if (batchFlushEnable) {
+			if (estimateSize != 0 && (estimateSize + mutationSize > batchFlushMaxSizeInBits || mutaionBuffer.size() + 1 > batchFlushMaxMutations)) {
+				synchronized (lock){
+					if (estimateSize != 0 && (estimateSize + mutationSize > batchFlushMaxSizeInBits
```

Comment:
Would you like to explain why not use HBase BufferedMutator?

Reply:
After discussion with Hequn Cheng in Jira ticket, I think we might be able to reduce some writes by pre computation with the mutator list.